From 1ed3b49650a757d17948140da9cc2fd75bca9fe8 Mon Sep 17 00:00:00 2001
Message-Id: <1ed3b49650a757d17948140da9cc2fd75bca9fe8.1658235456.git.maciek.borzecki@gmail.com>
In-Reply-To: <cover.1658235456.git.maciek.borzecki@gmail.com>
References: <cover.1658235456.git.maciek.borzecki@gmail.com>
From: John Johansen <john.johansen@canonical.com>
Date: Tue, 26 Jun 2018 20:19:19 -0700
Subject: [PATCH 3/5] UBUNTU: SAUCE: apparmor: fix use after free in
 sk_peer_label

Upstream-Status: Inappropriate [not author]
The patch was imported from Ubuntu Kernel git server
(https://git.launchpad.net/~ubuntu-kernel/ubuntu/+source/linux/+git/jammy) as
of commit id e9243f6a285589f49161faf0f96f4cf15c1dafae.

BugLink: http://bugs.launchpad.net/bugs/1778646
Signed-off-by: John Johansen <john.johansen@canonical.com>
Signed-off-by: Seth Forshee <seth.forshee@canonical.com>
Signed-off-by: Maciek Borzecki <maciek.borzecki@gmail.com>
---
 security/apparmor/lsm.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/security/apparmor/lsm.c b/security/apparmor/lsm.c
index 59a8dddab914a11b0b989695da937b2ca9cfc342..b1216ee7dd2f33fff6f9a6dd2f138fd34d43b3d5 100644
--- a/security/apparmor/lsm.c
+++ b/security/apparmor/lsm.c
@@ -1162,9 +1162,10 @@ static struct aa_label *sk_peer_label(struct sock *sk)
 {
 	struct sock *peer_sk;
 	struct aa_sk_ctx *ctx = SK_CTX(sk);
+	struct aa_label *label = ERR_PTR(-ENOPROTOOPT);
 
 	if (ctx->peer)
-		return ctx->peer;
+		return aa_get_label(ctx->peer);
 
 	if (sk->sk_family != PF_UNIX)
 		return ERR_PTR(-ENOPROTOOPT);
@@ -1172,14 +1173,15 @@ static struct aa_label *sk_peer_label(struct sock *sk)
 	/* check for sockpair peering which does not go through
 	 * security_unix_stream_connect
 	 */
-	peer_sk = unix_peer(sk);
+	peer_sk = unix_peer_get(sk);
 	if (peer_sk) {
 		ctx = SK_CTX(peer_sk);
 		if (ctx->label)
-			return ctx->label;
+			label = aa_get_label(ctx->label);
+		sock_put(peer_sk);
 	}
 
-	return ERR_PTR(-ENOPROTOOPT);
+	return label;
 }
 
 /**
@@ -1223,6 +1225,7 @@ static int apparmor_socket_getpeersec_stream(struct socket *sock,
 
 	}
 
+	aa_put_label(peer);
 done:
 	end_current_label_crit_section(label);
 
-- 
2.37.1

